src/cursor.c

```diff
@@ -600,6 +623,10 @@ constrain_cursor(struct server *server, struct wlr_pointer_constraint_v1
 	}
 	wl_list_remove(&seat->constraint_commit.link);
 	if (seat->current_constraint) {
+		if (!constraint) {
+			warp_cursor_to_constraint_hint(seat, seat->current_constraint);
+		}
```

Comment:
Does removing the condition and always executing `warp_cursor_to_constraint_hint()` fix it?
Currently we only warp the pointer if we reset the lock but not when we replace it with a new one.

Reply:
Nope, same behaviour.
Could it be the early bail for cursor_locked doing it?

Reply:
Okay, this may actually be a game bug; I've just done the same test on my game engine which warps the pointer to the center on unlock and it works fine in all cases there. Might test on Windows.

Reply:
Well, worth a shot.

> Could it be the early bail for cursor_locked doing it?
That is unlikely, that one is only triggered when actually moving the mouse and this PR should in fact update the cursor position (and image) even without movements.

If you are fine with the squashed commit and the current state I'd go for a merge of this PR.
We can always improve this in a new PR.

Reply:
> Well, worth a shot.
> 
> > Could it be the early bail for cursor_locked doing it?
> 
> That is unlikely, that one is only triggered when actually moving the mouse and this PR should in fact update the cursor position (and image) even without movements.
> 
> If you are fine with the squashed commit and the current state I'd go for a merge of this PR. We can always improve this in a new PR.

I just tested on X11 and it does center there even without movement. I'll look into this further but in it's current state I'm happy to have it merged.

Reply:
While thinking a bit about this, two potential reasons:
- it might be possible that the hint is only set when xwayland receives relative motion events which we only send if the cursor is actually moved. Thus no movement after creating the lock: no update to the cursor hint.
- it might be possible that it isn't committed at that point, e.g. the hint only being available via `constraint->pending.cursor_hint`.

Reply:
> While thinking a bit about this, two potential reasons:
> 
> * it might be possible that the hint is only set when xwayland receives relative motion events which we only send if the cursor is actually moved. Thus no movement after creating the lock: no update to the cursor hint.
> * it might be possible that it isn't committed at that point, e.g. the hint only being available via `constraint->pending.cursor_hint`.

These were also my thoughts; it's something tied to an event that isn't sent yet. It sure looks like it.